src/Sulu/Component/Content/Types/Block/ScheduleBlockVisitor.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Content\Types\Block;
+
+use Sulu\Bundle\HttpCacheBundle\CacheLifetime\CacheLifetimeRequestEnhancer;
+use Sulu\Component\Content\Compat\Block\BlockPropertyType;
+use Sulu\Component\Webspace\Analyzer\RequestAnalyzerInterface;
+
+class ScheduleBlockVisitor implements BlockVisitorInterface
+{
+    /**
+     * @var RequestAnalyzerInterface
+     */
+    private $requestAnalyzer;
+
+    /**
+     * @var CacheLifetimeRequestEnhancer
+     */
+    private $cacheLifetimeRequestEnhancer;
+
+    public function __construct(
+        RequestAnalyzerInterface $requestAnalyzer,
+        CacheLifetimeRequestEnhancer $cacheLifetimeRequestEnhancer
+    ) {
+        $this->requestAnalyzer = $requestAnalyzer;
+        $this->cacheLifetimeRequestEnhancer = $cacheLifetimeRequestEnhancer;
+    }
+
+    public function visit(BlockPropertyType $block): ?BlockPropertyType
+    {
+        $blockPropertyTypeSettings = $block->getSettings();
+
+        if (!\is_array($blockPropertyTypeSettings)
+            || !isset($blockPropertyTypeSettings['schedules_enabled'])
+            || !$blockPropertyTypeSettings['schedules_enabled']
+            || !isset($blockPropertyTypeSettings['schedules'])
+            || !\is_array($blockPropertyTypeSettings['schedules'])
+        ) {
+            return $block;
+        }
+
+        $now = $this->requestAnalyzer->getDateTime();
+        $nowTimestamp = $now->getTimestamp();
+
+        $returnBlock = false;
+
+        foreach ($blockPropertyTypeSettings['schedules'] as $schedule) {
+            switch ($schedule['type']) {
+                case 'fixed':
+                    $start = new \DateTime($schedule['start']);
+                    $end = new \DateTime($schedule['end']);
+
+                    $this->cacheLifetimeRequestEnhancer->setCacheLifetime($start->getTimestamp() - $nowTimestamp);
+                    $this->cacheLifetimeRequestEnhancer->setCacheLifetime($end->getTimestamp() - $nowTimestamp);
+
+                    if ($now >= $start && $now <= $end) {
+                        $returnBlock = true;
+                        continue 2;
+                    }
+                    break;
+                case 'weekly':
+                    $year = $now->format('Y');
+                    $month = $now->format('m');
+                    $day = $now->format('d');
+
+                    $start = new \DateTime($schedule['start']);
+                    $start->setDate($year, $month, $day);
+                    $end = new \DateTime($schedule['end']);
+                    $end->setDate($year, $month, $day);
+
+                    if ($end < $start) {
+                        $start->modify('-1 day');
```

Comment:
shouldnt this be `$end->modify('+1 day')`? 🤔

Reply:
No, then it does not work if you e.g. set the schedule from friday 20:00 to saturday 04:00 and you visit on saturday 03:00, because the start date is used to check if the selected week days are matching.